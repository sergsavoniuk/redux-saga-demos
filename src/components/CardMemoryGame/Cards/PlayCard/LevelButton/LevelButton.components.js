import styled from 'styled-components';

export const Button = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 0;
  border: none;
  color: #fff;
  font-family: 'Open Sans';
  font-size: 0.6em;
  font-weight: bold;
  background-color: transparent;
  text-transform: uppercase;
  text-decoration: none;

  &:hover {
    background-color: #ffffff1a;
  }
`;
