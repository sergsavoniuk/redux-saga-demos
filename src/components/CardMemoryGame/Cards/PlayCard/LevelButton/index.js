export { default, LevelButton } from './LevelButton';
