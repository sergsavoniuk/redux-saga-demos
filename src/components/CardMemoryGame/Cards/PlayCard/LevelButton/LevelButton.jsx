import React from 'react';
import { string, func } from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { Button } from './LevelButton.components';
import { ActionCreators } from 'redux/cardGame';
import { ROUTES } from 'constants/routes';

export function LevelButton({ level, chooseLevel }) {
  function handleChooseLevel() {
    chooseLevel(level);
  }

  return (
    <Button
      as={Link}
      to={{
        pathname: `${ROUTES.CardGameApp.Play}`,
        search: `?level=${level}`,
      }}
      onClick={handleChooseLevel}
    >
      {level}
    </Button>
  );
}

LevelButton.propTypes = {
  level: string.isRequired,
  chooseLevel: func.isRequired,
};

export default connect(
  null,
  {
    chooseLevel: ActionCreators.chooseLevel,
  },
)(LevelButton);
