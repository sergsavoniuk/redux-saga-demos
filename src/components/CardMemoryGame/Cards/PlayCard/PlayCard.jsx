import React from 'react';

import withMemo from 'utils/withMemo';
import { Card } from '../Cards.components';
import { LEVELS } from 'constants/cardGame/levels';
import { CardPropTypes } from '../Card';

import LevelButton from './LevelButton';

const { Casual, Medium, Hard } = LEVELS;

function PlayCard({ name, isFlipped, onCardClick }) {
  const content = !isFlipped ? (
    <p>{name}</p>
  ) : (
    <>
      <LevelButton level={Casual} />
      <LevelButton level={Medium} />
      <LevelButton level={Hard} />
    </>
  );
  return (
    <Card fixed={isFlipped} onClick={onCardClick}>
      {content}
    </Card>
  );
}

PlayCard.propTypes = CardPropTypes;

export default withMemo(PlayCard, ['isFlipped']);
