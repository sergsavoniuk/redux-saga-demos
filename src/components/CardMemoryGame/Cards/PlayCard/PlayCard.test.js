import React from 'react';
import { shallow } from 'enzyme';

import PlayCard from './PlayCard';
import LevelButton from './LevelButton';

describe('<PlayCard />', () => {
  const onCardClick = jest.fn();

  function setup(props) {
    const defaultProps = {
      name: 'P',
      isFlipped: false,
      onCardClick,
    };

    const wrapper = shallow(<PlayCard {...defaultProps} {...props} />);

    return {
      wrapper,
      props: defaultProps,
    };
  }

  afterEach(() => {
    onCardClick.mockClear();
  });

  test('renders correctly unflipped card', () => {
    const { wrapper, props } = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.text()).toMatch(props.name);
  });

  test('renders correctly flipped card', () => {
    const { wrapper } = setup({ isFlipped: true });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(LevelButton)).toHaveLength(3);
  });

  test('should flip a card after a click', () => {
    const { wrapper, props } = setup();

    expect(wrapper.text()).toMatch(props.name);

    wrapper.simulate('click');
    expect(onCardClick).toHaveBeenCalledTimes(1);

    wrapper.setProps({ isFlipped: true });
    expect(wrapper.find(LevelButton)).toHaveLength(3);

    wrapper.simulate('click');
    expect(onCardClick).toHaveBeenCalledTimes(2);

    wrapper.setProps({ isFlipped: false });
    expect(wrapper.text()).toMatch(props.name);
  });
});
