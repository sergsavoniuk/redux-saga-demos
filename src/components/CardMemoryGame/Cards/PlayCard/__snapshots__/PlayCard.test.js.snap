// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayCard /> renders correctly flipped card 1`] = `
<styled.div
  fixed={true}
  onClick={[MockFunction]}
>
  <ConnectFunction
    level="casual"
  />
  <ConnectFunction
    level="medium"
  />
  <ConnectFunction
    level="hard"
  />
</styled.div>
`;

exports[`<PlayCard /> renders correctly unflipped card 1`] = `
<styled.div
  fixed={false}
  onClick={[MockFunction]}
>
  <p>
    P
  </p>
</styled.div>
`;
